### DISCUSSION TITLE
Data is not persistent on Heroku? [HELP]

### DISCUSSION THREAD

**Discussion by madrackwp** [python-telegram-bot/python-telegram-bot#2600] [Q&A]
I created a simple assignment tracker bot to use. During the coding and testing of my bot, I ran everything locally on my computer's terminal and data seems to be persistent. However, after deploying it on Heroku, data is no longer saved.

For reference, I use context.user_data[] to save all the data for my bot. I was wondering if I missed out something?

**Comment by Bibo-Joshi** [comment]
Please have a look at https://github.com/python-telegram-bot/python-telegram-bot/wiki/Making-your-bot-persistent
AFAIK heroku clears stored files every once in a while, so `PicklePersistence` is probably not the best choice, but you might be interested in https://github.com/python-telegram-bot/ptbcontrib/tree/main/ptbcontrib/postgres_persistence

**Reply by madrackwp** [reply]
Thanks for the reply. But I may have come up with a simpler work around to this problem. I can simply pin the data into the chat and refer to that whenever I restart the bot. Problem is, I'm having issues figuring out how to simply pin messages. I was wondering if you could help me with it?

**Reply by Poolitzer** [reply]
If the bot is an admin with pin rights (or is in a private chat), you can simply use `update.effective_message.pin`. You need to be aware that you only get the most recent pinned message with get_chat, so if a user pins another message (or unpins the one from the bot), your code breaks.
